README.md
[![Gem Version](https://badge.fury.io/rb/float-formats.svg)](http://badge.fury.io/rb/float-formats)
[![Build Status](https://travis-ci.org/jgoizueta/float-formats.svg)](https://travis-ci.org/jgoizueta/float-formats)

# Introduction

Float-Formats is a Ruby package with methods to handle diverse floating-point formats.
These are some of the things that can be done with it:

* Encoding and decoding numerical values in specific floating point representations.
* Conversion of floating-point data between different formats.
* Obtaining properties of floating-point formats (ranges, precision, etc.)
* Exploring and learning about floating point representations.
* Definition and testing of new floating-point formats.

# Installation

To install the gem manually:

    gem install float-formats

You can find the code in GitHub:

* http://github.com/jgoizueta/float-formats/

# Predefined formats

A number of common formats are defined as constants in the Flt module:

## IEEE 754-2008

**Binary** floating point representations in little endian order:

* IEEE_binary16 (half precision),
* IEEE_binary32 (single precision),
* IEEE_binary64 (double precision),
* IEEE_binary80 (extended), IEEE_binary128 (quadruple precision) and
  as little endian: IEEE_binary16_BE, etc.

**Decimal** formats (using DPD):

* IEEE_decimal32, IEEE_decimal64 and IEEE_decimal128.

**Interchange binary & decimal** formats:

* IEEE_binary256, IEEE_binary512, IEEE_binary1024, IEEE_decimal192, IEEE_decimal256.

Others can be defined with IEEE.interchange_binary and IEEE.interchange_decimal
(see the IEEE module).

## Legacy

Formats of historical interest, some of which are found
in file formats still in use.

**Mainframe/supercomputer** formats:
Univac 1100 (UNIVAC_SINGLE, UNIVAC_DOUBLE),
IBM 360 etc. (IBM32, IBM64 and IBM128),
CDC 6600/7600: (CDC_SINGLE, CDC_DOUBLE),
Cray-1: (CRAY).

**Minis**: PDP11 and Vaxes: (PDP11_F, PDP11_D, VAX_F, VAX_D, VAX_G and VAX_H),
HP3000: (XS256, XS256_DOUBLE),
Wang 2200: (WANG2200).

**Microcomputers** (software implementations):
Apple II: (APPLE),
Microsoft Basic, Spectrum, etc.: (XS128),
Microsoft Quickbasic: (MBF_SINGLE, MBF_DOUBLE),
Borland Pascal: (BORLAND48).

**Embedded systems**:
Formats used in the Intel 8051 by the C51 compiler:
(C51_BCD_FLOAT, C51_BCD_DOUBLE and C51_BCD_LONG_DOUBLE).

**Minifloats**:
AI Formats: (BFLOAT16, MSFP8, MSFP9, MSFP10, MSFP11)
Khronos Vulkan unsigned formats: (KHRONOS_VULKAN_UNSIGNED10, KHRONOS_VULKAN_UNSIGNED11)

## Calculators

Formats used in HP RPL calculators: (RPL, RPL_X),
HP-71B formats (HP71B, HP71B_X)
and classic HP 10 digit calculators: (HP_CLASSIC).

# Using the pre-defined formats

    require 'rubygems'
    require 'float-formats'
    include Flt

The properties of the floating point formats can be queried (which can be
used for tables or reports comparing different formats):

Size in bits of the representations:

    puts IEEE_binary32.total_bits                        # -> 32

Numeric radix:

    puts IEEE_binary32.radix                             # -> 2

Digits of precision (radix-based)

    puts IEEE_binary32.significand_digits                # -> 24

Minimum and maximum values of the radix-based exponent:

    puts IEEE_binary32.radix_min_exp                     # -> -126
    puts IEEE_binary32.radix_max_exp                     # -> 127

Decimal precision

    puts IEEE_binary32.decimal_digits_stored             # -> 6
    puts IEEE_binary32.decimal_digits_necessary          # -> 9

Minimum and maximum decimal exponents:

    puts IEEE_binary32.decimal_min_exp                   # -> -37
    puts IEEE_binary32.decimal_max_exp                   # -> 38

## Encode and decode numbers

For each floating-point format class there is a constructor method with the same
name which can build a floating-point value from a variety of parameters:

* Using three integers:
  the sign (+1 for +, -1 for -), the significand (coefficient or mantissa)
  and the exponent.
* From a text numeral (with an optional `Numerals` format specifier)
* From a number : converts a numerical value
  to a floating point representation.

Examples:

    File.open('binary_file.dat','wb'){|f| f.write IEEE_binary80('0.1').to_bytes}
    puts IEEE_binary80('0.1').to_hex(true)           # -> CD CC CC CC CC CC CC CC FB 3F
    puts IEEE_binary80(0.1).to_hex(true)             # -> CD CC CC CC CC CC CC CC FB 3F
    puts IEEE_binary80(+1,123,-2).to_hex(true)       # -> 00 00 00 00 00 00 00 F6 03 40
    puts IEEE_decimal32('1.234').to_hex(true)        # -> 22 20 05 34

A floating-point encoded value can be converted to useful formats with the to_ and similar methods:

* `split` (split as integral sign, significand, exponent)
* `to_text`
* `to(num_class)`

Examples:

    v = IEEE_binary80.from_bytes(File.read('binary_file.dat'))
    puts v.to(Rational)                              # -> 1/10
    puts v.split.inspect                             # -> [1, 14757395258967641293, -67]
    puts v.to_text                                   # -> 0.1
    puts v.to(Float)                                 # -> 0.1
    puts v.to_hex                                    # -> CDCCCCCCCCCCCCCCFB3F
    puts v.to_bits                                   # -> 00111111111110111100110011001100110011001100110011001100110011001100110011001101
    puts v.to_bits_text(16)                          # -> 3ffbcccccccccccccccd

## Special values:

Let's show the decimal expression of some interesting values using
3 significative digits:

    fmt = Numerals::Format[mode: :general, rounding: [precision: 3]]
    puts IEEE_SINGLE.min_value.to_text(fmt)             # -> 1e-45
    puts IEEE_SINGLE.min_normalized_value.to_text(fmt)  # -> 1.18e-38
    puts IEEE_SINGLE.max_value.to_text(fmt)             # -> 3.40e38
    puts IEEE_SINGLE.epsilon.to_text(fmt)               # -> 1.19e-7

## Convert between formats

    v = IEEE_EXTENDED.from_text('1.1')
    v = v.convert_to(IEEE_SINGLE)
    v = v.convert_to(IEEE_DEC64)

# Tools for the native floating point format

This is an optional module to perform conversions and manipulate the native Float format.

    require 'float-formats/native'
    include Flt

    puts float_shortest_dec(0.1)                     # -> 0.1
    puts float_significant_dec(0.1)                  # -> 0.10000000000000001
    puts float_dec(0.1)                              # -> 0.1000000000000000055511151231257827021181583404541015625

History.txt
== 0.4.1 2020-04-07

* Development dependencies updated

== 0.4.0 2020-04-07

* New formats: BFLOAT16, MSFP8, MSFP9, MSFP10, MSFP11, KHRONOS_VULKAN_UNSIGNED10, KHRONOS_VULKAN_UNSIGNED11
* New aliases: MBF_EXTENDEND, HFP_SINGLE, HFP_DOUBLE, HFP_EXTENDED
* Development dependencies updated

== 0.3.0 2015-04-05

* Nio dependency replaced by Numerals
* Ruby 1.8 compatibility dropped

== 0.2.1 2015-03-31

* Compatibility issues with Ruby 1.9/2

== 0.2.0 2009-08-06

*  New redefined interface
*  The sign in split form (sign, significand, exponent) now
   have the sign as an integer +1,-1 rather than using the
   internal representation (0, 1, etc.)
*  Arithmetic is possible on FP values
*  Requires Flt 1.0.0 and Nio 0.2.4

== 0.1.1 2007-11-15

*  HP-71B formats defined
*  Add half precision IEEE format (binary16)
*  New names for IEEE formats
*  Add some IEEE 754r interchange formats
*  new methods hex_to_float, hex_from_float in float-formats/native
*  Allow non-bcd values in fields of BCD formats by passing
   hex values as Strings; allow such values to be used for
   nan/infinity exponents.
*  Nio 0.2.1 is now required
*  Handle special values (Infinities and NaN) in #from_fmt, #from_number
*  Add ulp methods to Value and FP classes and to Float
*  Bug fixes
    - Fix the encoding-decoding of nan and infinity in Decimal format.
    - Fix the decoding of NaN in Binary & Hexadecimal
    - The definition of IEEE_binary128 was not correct

== 0.1.0 2007-11-04

* Initial release


lib/float-formats/formats.rb
# Float-Formats
# Definition of common floating-point formats

require 'enumerator'

require 'float-formats/classes.rb'


module Flt


# Floating Point Format Definitions ==========================================

# Helper methods to define IEEE 754r formats
module IEEE
  # Define an IEEE binary format by passing parameters in a hash;
  # :significand and :exponent are used to defined the fields,
  # optional parameters may follow.
  def self.binary(name, parameters)
    significand_bits = parameters[:significand]
    exponent_bits = parameters[:exponent]
    Flt.define(name, {
      base: BinaryFormat,
      fields: [:significand, significand_bits, :exponent, exponent_bits, :sign, 1],
      bias: 2**(exponent_bits-1)-1, bias_mode: :scientific_significand,
      hidden_bit: true,
      endianness: :little_endian,
      round: :half_even,
      gradual_underflow: true,
      infinity: true, nan: true
      }.merge(parameters)
    )
  end

  # Define an IEEE binary interchange format given its width in bits
  def self.interchange_binary(name,width_in_bits, options={})
    unless (width_in_bits%32)==0 && (width_in_bits/32)>=4
      raise "Invalid IEEE binary interchange format definition: size (#{width_in_bits}) is not valid"
    end
    p = width_in_bits - (4*Math.log(width_in_bits)/Math.log(2)).round.to_i + 13
    binary(name, { significand: p-1, exponent: width_in_bits-p }.merge(options))
  end

  # Define an IEEE decimal format by passing parameters in a hash;
  # :significand and :exponent are used to defined the fields,
  # optional parameters may follow.
  def self.decimal(name,parameters)
    significand_continuation_bits = parameters[:significand]
    exponent_continuation_bits = parameters[:exponent]
    Flt.define(name, {
      base: DPDFormat,
      fields: [:significand_continuation, significand_continuation_bits, :exponent_continuation, exponent_continuation_bits, :combination, 5, :sign, 1],
      normalized: false,
      endianness: :big_endian,
      gradual_underflow: true,
      infinity: true, nan: true,
      round: :half_even
    }.merge(parameters))
  end

  # Define an IEEE decimal interchange format given its width in bits
  def self.interchange_decimal(name,width_in_bits, options={})
    raise "Invalid IEEE decimal interchange format definition: size (#{width_in_bits}) is not valid" unless (width_in_bits%32)==0
    p = width_in_bits*9/32 - 2
    t = (p-1)*10/3
    w = width_in_bits - t - 6
    decimal(name,{ significand: t, exponent: w }.merge(options))
  end

end

# IEEE 754 binary types, as stored in little endian architectures such as Intel, Alpha

IEEE.binary :IEEE_binary16, significand: 10, exponent: 5
IEEE.binary :IEEE_binary32, significand: 23,exponent: 8
IEEE.binary :IEEE_binary64, significand: 52,exponent: 11
IEEE.binary :IEEE_binary80, significand: 64,exponent: 15, hidden_bit: false, min_encoded_exp: 1
IEEE.binary :IEEE_binary128, significand: 112,exponent: 15

# IEEE 754 in big endian order (SPARC, Motorola 68k, PowerPC)

IEEE.binary :IEEE_binary16_BE, significand: 10, exponent: 5, endianness: :big_endian
IEEE.binary :IEEE_binary32_BE, significand: 23,exponent: 8, endianness: :big_endian
IEEE.binary :IEEE_binary64_BE, significand: 52,exponent: 11, endianness: :big_endian
IEEE.binary :IEEE_binary80_BE, significand: 64,exponent: 15, endianness: :big_endian, hidden_bit: false, min_encoded_exp: 1
IEEE.binary :IEEE_binary128_BE, significand: 112,exponent: 15, endianness: :big_endian


# some IEEE745r interchange binary formats

IEEE.interchange_binary :IEEE_binary256, 256
IEEE.interchange_binary :IEEE_binary512, 512
IEEE.interchange_binary :IEEE_binary1024, 1024
IEEE.interchange_binary :IEEE_binary256_BE, 256, endianness: :big_endian
IEEE.interchange_binary :IEEE_binary512_BE, 512, endianness: :big_endian
IEEE.interchange_binary :IEEE_binary1024_BE, 1024, endianness: :big_endian

# old names
IEEE_binaryx = IEEE_binary80
IEEE_HALF = IEEE_binary16
IEEE_SINGLE = IEEE_binary32
IEEE_DOUBLE = IEEE_binary64
IEEE_EXTENDED = IEEE_binary80
IEEE_QUAD = IEEE_binary128
IEEE_128 = IEEE_binary128
IEEE_H_BE = IEEE_binary16_BE
IEEE_S_BE = IEEE_binary32_BE
IEEE_D_BE = IEEE_binary64_BE
IEEE_X_BE = IEEE_binary80_BE
IEEE_128_BE = IEEE_binary128_BE
IEEE_Q_BE = IEEE_binary128_BE


# Decimal IEEE 754r formats

IEEE.decimal :IEEE_decimal32, significand: 20, exponent: 6
IEEE.decimal :IEEE_decimal64, significand: 50, exponent: 8
IEEE.decimal :IEEE_decimal128, significand: 110, exponent: 12

# some IEEE745r interchange binary formats

IEEE.interchange_decimal :IEEE_decimal96, 96
IEEE.interchange_decimal :IEEE_decimal192, 192
IEEE.interchange_decimal :IEEE_decimal256, 256

# old names

IEEE_DEC32 = IEEE_decimal32
IEEE_DEC64 = IEEE_decimal64
IEEE_DEC128 = IEEE_decimal128

# Brain Floating Point (bfloat16) - Google Brain 16-bit truncated version of IEEE_binary32
# Google Cloud TPUs, TensorFlow
# Intel AI processors: Nervana NNP-L1000, Xeon processors (AVX-512 BF16 extensions)
# Intel FPGAs
# ARMv8.6-A
# AMD ROCm libraries

IEEE.binary :BFLOAT16, significand: 7, exponent: 8, endianness: :little_endian
IEEE.binary :BFLOAT16_BE, significand: 7, exponent: 8, endianness: :big_endian

# MSFP8-11 Microsoft Floating Point (Project Brainwave)

IEEE.binary :MSFP8, significand: 2, exponent: 5, endianness: :little_endian
IEEE.binary :MSFP9, significand: 3, exponent: 5, endianness: :little_endian
IEEE.binary :MSFP10, significand: 4, exponent: 5, endianness: :little_endian
IEEE.binary :MSFP11, significand: 5, exponent: 5, endianness: :little_endian
IEEE.binary :MSFP8_BE, significand: 2, exponent: 5, endianness: :big_endian
IEEE.binary :MSFP9_BE, significand: 3, exponent: 5, endianness: :big_endian
IEEE.binary :MSFP10_BE, significand: 4, exponent: 5, endianness: :big_endian
IEEE.binary :MSFP11_BE, significand: 5, exponent: 5, endianness: :big_endian

# Khronos Vulkan unsigned minifloat formats

Flt.define BinaryFormat, :KHRONOS_VULKAN_UNSIGNED11,
  fields:[:significand, 6, :exponent, 5], # :sign 1
  bias: 2**(5-1)-1, bias_mode: :scientific_significand,
  hidden_bit: true,
  endianness: :little_endian,
  round: :half_even,
  gradual_underflow: true,
  infinity: true, nan: true

Flt.define BinaryFormat, :KHRONOS_VULKAN_UNSIGNED10,
  fields:[:significand, 5, :exponent, 5], # :sign 1
  bias: 2**(5-1)-1, bias_mode: :scientific_significand,
  hidden_bit: true,
  endianness: :little_endian,
  round: :half_even,
  gradual_underflow: true,
  infinity: true, nan: true

# Excess 128 used by Microsoft Basic in 8-bit micros, Spectrum, ...

Flt.define BinaryFormat, :XS128,
  fields:[:significand,31,:sign,1,:exponent,8],
  bias: 128, bias_mode: :fractional_significand,
  hidden_bit: true,
  endianness: :big_endian, round: :half_up,
  gradual_underflow: false, infinity: false, nan: false

# HP-3000 excess 256 format, HP-Tandem...

Flt.define BinaryFormat, :XS256,
  fields:[:significand,22,:exponent,9,:sign,1],
  bias: 256, bias_mode: :scientific_significand,
  hidden_bit: true, min_encoded_exp: 0,
  endianness: :big_endian, round: :half_up,
  gradual_underflow: false, infinity: false, nan: false

Flt.define :XS256_DOUBLE, BinaryFormat,
  fields:[:significand,54,:exponent,9,:sign,1],
  bias: 256, bias_mode: :scientific_significand,
  hidden_bit: true, min_encoded_exp: 0,
  endianness: :big_endian, round: :half_up,
  gradual_underflow: false, infinity: false, nan: false

# Borland Pascal 48 bits "Real" Format

Flt.define :BORLAND48, BinaryFormat,
  fields:[:exponent,8,:significand,39,:sign,1],
  bias: 128, bias_mode: :fractional_significand,
  hidden_bit: true,
  endianness: :little_endian,
  gradual_underflow: false, infinity: false, nan: false

# Microsoft Binary Floating-point (Quickbasic)

Flt.define :MBF_SINGLE, BinaryFormat,
  fields:[:significand,23,:sign,1,:exponent,8],
  bias: 128, bias_mode: :fractional_significand,
  hidden_bit: true,
  endianness: :little_endian,
  gradual_underflow: false, infinity: false, nan: false

Flt.define :MBF_DOUBLE, BinaryFormat,
  fields:[:significand,55,:sign,1,:exponent,8],
  bias: 128, bias_mode: :fractional_significand,
  hidden_bit: true,
  endianness: :little_endian,
  gradual_underflow: false, infinity: false, nan: false

MBF_EXTENDEND = XS128

# DEC formats (VAX)

Flt.define :VAX_F, BinaryFormat,
  fields:[:significand, 23, :exponent, 8, :sign, 1],
  bias: 128, bias_mode: :fractional_significand,
  hidden_bit: true,
  endianness: :little_big_endian,
  gradual_underflow: false, infinity: false, nan: false

Flt.define :VAX_D, BinaryFormat,
  fields:[:significand, 55, :exponent, 8, :sign, 1],
  bias: 128, bias_mode: :fractional_significand,
  hidden_bit: true,
  endianness: :little_big_endian,
  gradual_underflow: false, infinity: false, nan: false

Flt.define :VAX_G, BinaryFormat,
  fields:[:significand, 52, :exponent, 11, :sign, 1],
  bias: 1024, bias_mode: :fractional_significand,
  hidden_bit: true,
  endianness: :little_big_endian,
  gradual_underflow: false, infinity: false, nan: false

Flt.define :VAX_H, BinaryFormat,
  fields:[:significand, 112, :exponent, 15, :sign, 1],
  bias: 16384, bias_mode: :fractional_significand,
  hidden_bit: true,
  endianness: :little_big_endian,
  gradual_underflow: false, infinity: false, nan: false

# DEC PDP 11 variants (minimum exponent used for normalized values other than zero)

Flt.define :PDP11_F, BinaryFormat,
  fields:[:significand, 23, :exponent, 8, :sign, 1],
  bias: 128, bias_mode: :fractional_significand,
  hidden_bit: true, min_encoded_exp: 0,
  endianness: :little_big_endian,
  gradual_underflow: false, infinity: false, nan: false

Flt.define :PDP11_D, BinaryFormat,
  fields:[:significand, 55, :exponent, 8, :sign, 1],
  bias: 128, bias_mode: :fractional_significand,
  hidden_bit: true, min_encoded_exp: 0,
  endianness: :little_big_endian,
  gradual_underflow: false, infinity: false, nan: false


# Format used in HP Saturn-based RPL calculators (HP48,HP49,HP50, also HP32s, HP42s --which use RPL internally)
# (these formats are not used in the HP-71B which is a Saturn, non-RPL machine)

Flt.define :SATURN, BCDFormat,
  fields:[:prolog,5,:exponent,3,:significand,12,:sign,1],
  fields_handler: lambda{|fields| fields[0]=2933},
  exponent_mode: :radix_complement,
  endianness: :little_endian, round: :half_even,
  gradual_underflow: false, infinity: false, nan: false

Flt.define :SATURN_X, BCDFormat,
  fields:[:prolog,5,:exponent,5,:significand,15,:sign,1],
  fields_handler: lambda{|fields| fields[0]=2955},
  exponent_mode: :radix_complement,
  endianness: :little_endian, round: :half_even,
  gradual_underflow: false, infinity: false, nan: false


RPL = SATURN
RPL_X = SATURN_X

# SATURN HP-71B (IEEE, NON-RPL) formats

# HP-71B REAL format (12-form) which is stored in a single register
Flt.define :HP71B, BCDFormat,
  fields:[:exponent,3,:significand,12,:sign,1],
  exponent_mode: :radix_complement,
  endianness: :little_endian, round: :half_even,
  gradual_underflow: true, infinity: true, nan: true,
  denormal_encoded_exp: 501,
  nan_encoded_exp:"F01", # signaling NaN is F02
  infinite_encoded_exp:"F00"

# HP-71B internal 15-digit format (15-form), stored in a pair of registers
# we use here a little-endian order for the registers, otherwise the
# definition would be [:significand,15,:unused1,1,:exponent,5,:unused2,10,:sign,1]
Flt.define :HP71B_X, BCDFormat,
  fields:[:exponent,5,:unused2,10,:sign,1, :significand,15,:unused1,1],
  exponent_mode: :radix_complement,
  endianness: :little_endian, round: :half_even,
  gradual_underflow: false, infinity: true, nan: true,
  nan_encoded_exp:"00F01",
  infinite_encoded_exp:"00F00"

# Format used in classic HP calculators (HP-35, ... HP-15C)
# Endianness is indeterminate, since these machines have named registers that
# hold a floating-point value in a single 56-bit word.
# (But intra-word field/nibble addressing is little-endian)
Flt.define :HP_CLASSIC, BCDFormat,
  fields:[:exponent,3,:significand,10,:sign,1],
  exponent_mode: :radix_complement,
  min_exp:-99, max_exp: 99, # the most significant nibble of exponent if for sign only
  endianness: :big_endian, round: :half_up,
  gradual_underflow: false, infinity: false, nan: false


# IBM Floating Point Architecture (IBM 360,370, DG Eclipse, ...)

# short
Flt.define :IBM32, HexadecimalFormat,
  fields:[:significand,24,:exponent,7,:sign,1],
  bias: 64, bias_mode: :fractional_significand,
  endianness: :big_endian

# long
Flt.define :IBM64, HexadecimalFormat,
  fields:[:significand,56,:exponent,7,:sign,1],
  bias: 64, bias_mode: :fractional_significand,
  endianness: :big_endian

# extended: two long values pasted together
Flt.define :IBM128, HexadecimalFormat,
  fields:[:significand,14*4,:lo_exponent,7,:lo_sign,1,:significand,14*4,:exponent,7,:sign,1],
  fields_handler: lambda{|fields| fields[1]=(fields[4]>=14&&fields[4]<127) ? fields[4]-14 : fields[4];fields[2]=fields[5] },
  bias: 64, bias_mode: :fractional_significand,
  min_encoded_exp: 14, # to avoid out-of-range exponents in the lower half
  endianness: :big_endian

# It think this has never been used:
Flt.define :IBMX, HexadecimalFormat,
  fields:[:significand,14*4,:exponent,7,:unused_sign,1,:significand,14*4,:exponent,7,:sign,1],
  fields_handler: lambda{|fields| fields[2]=0},
  bias: 8192, bias_mode: :fractional_significand,
  endianness: :big_endian

HFP_SINGLE = IBM32
HFP_DOUBLE = IBM64
HFP_EXTENDED = IBM128

# Cray-1
Flt.define :CRAY, BinaryFormat,
  fields:[:significand,48,:exponent,15,:sign,1],
  bias: 16384, bias_mode: :fractional_significand,
  hidden_bit: false,
  min_encoded_exp: 8192, max_encoded_exp: 24575, zero_encoded_exp: 0,
  endianness: :big_endian,
  gradual_underflow: false, infinity: false, nan: false

# CDC 6600/7600
# Byte endianness is arbitrary, since these machines were word-addressable, in 60-bit words,
# but the two words used in double precision numbers are in big endian order
# TO DO: apply this:
# The exponent encoded value 1023 is used for NaN
# The exponent encoded value    0 is used for underflow
# The exponent encoded value 2047 is used for overflow
#
# The exponent needs special treatment, because instead of excess encoding, which is equivalent to two's complement followed
# by sign bit reversal, one's complement followed by sign bit reversal is used, which is equivalent
# to use a bias diminished by one for negative exponents. Note that the exponent encoded value that equals the bias is
# not used (is used as a NaN indicator)
class CDCFormat < BinaryFormat # :nodoc:
  def self.encode_exponent(e,mode)
    ee = super
    ee -= 1 if e<0
    ee
  end
  def self.decode_exponent(ee,mode)
    e = super
    e += 1 if e<0
    e
  end
end

Flt.define :CDC_SINGLE, CDCFormat,
  fields:[:significand,48, :exponent,11, :sign,1],
  bias: 1024, bias_mode: :integral_significand,
  min_exp:-1023,
  neg_mode: :diminished_radix_complement,
  hidden_bit: false,
  endianess: :big_endian,
  gradual_underflow: false, infinity: false, nan: false

# The CDC_DOUBLE can be splitted in two CDC_SINGLE values:
#   get_bitfields(v,[CDC_SINGLE.total_bits]*2,CDC_DOUBLE.endianness).collect{|x| int_to_bytes(x,0,CDC_SINGLE.endianness)}
# and the value of the double is the sum of the values of the singles.
# Unlike the single, we must use :fractional_significand mode because with :integral_significand
# the exponent would refer to the whole significand, but it must refer only to the most significant half.
# we substract the number of bits in the single to the bias and exponent because of this change,
# and add 48 to the min_exponent to avoid the exponent of the low order single to be out of range
# because the exponent of the low order single is adjusted to
# the position of its digits by substracting 48 from the high order exponent
# when its exponent would be out of range
# Note that when computing the low order exponent with the fields handler we must take into account the sign
# because for negative numbers all the fields are one-complemented.
Flt.define :CDC_DOUBLE, CDCFormat,
  fields:[:significand,48,:lo_exponent,11,:lo_sign,1,:significand,48,:exponent,11,:sign,1],
  fields_handler: lambda{|fields|
        fields[1]=(fields[4]>0&&fields[4]<2047) ? fields[4]-((-1)**fields[5])*48 : fields[4]
        fields[2]=fields[5]
  },
  bias: 1024-48, bias_mode: :fractional_significand,
  min_encoded_exp: 48+1, # + 1 because the bias for negative is 1023
  neg_mode: :diminished_radix_complement,
  hidden_bit: false,
  endianess: :big_endian,
  gradual_underflow: false, infinity: false, nan: false


# Univac 1100
# Byte endianness is arbitrary, since these machines were word-addressable, in 36-bit words,
# but the two words used in double precision numbers are in big endian order
Flt.define :UNIVAC_SINGLE, BinaryFormat,
  fields:[:significand,27, :exponent,8, :sign,1],
  bias: 128, bias_mode: :fractional_significand,
  neg_mode: :diminished_radix_complement,
  hidden_bit: false,
  endianess: :big_endian,
  gradual_underflow: false, infinity: false, nan: false

Flt.define :UNIVAC_DOUBLE, BinaryFormat,
  fields:[:significand,60, :exponent,11, :sign,1],
  bias: 1024, bias_mode: :fractional_significand,
  neg_mode: :diminished_radix_complement,
  hidden_bit: false,
  endianess: :big_endian,
  gradual_underflow: false, infinity: false, nan: false


# :stopdoc: # the next definition is not handled correctly by RDoc
Flt.define(:APPLE_INSANE,BinaryFormat,
  fields:[:significand,23,:sign,1,:exponent,8],
  bias: 128, bias_mode: :scientific_significand,
  hidden_bit: false, min_encoded_exp: 0,
  neg_mode: :radix_complement_significand,
  endianness: :big_endian,
  gradual_underflow: true, infinity: false, nan: false) { |fp|
  # This needs a peculiar treatment for the negative values, which not simply use two's complement
  # but also avoid having the sign and msb of the significand equal.
  # Note that here we have a separate sign bit, but it can also be considered as the msb of the significand
  # in two's complement, in which case the radix point is after the two msbs, which are the ones that
  # should not be equal. (except for zero and other values using the minimum exponent).
  def fp.neg_significand_exponent(s,f,e)
    #puts "before: #{s} #{f.to_s(16)}"
    f,e = super
    s = (s+1)%2
    #print "minus #{f.to_s(16)}"
    if e>@zero_encoded_exp && f==0
      f = 1<<(significand_digits-1)
      e += 1
    else
      while (e>@zero_encoded_exp) && (f>>(significand_digits-1))&1 == s
        e -= 1
        f = (f << 1) & (radix_power(significand_digits)-1)
        #print " << "
      end
    end
    #puts ""
    [f,e]
  end
}
# :startdoc:


# Sofware floating point implementatin for the Apple II (6502)
# the significand & sign are a single field in two's commplement
APPLE = APPLE_INSANE

# Wang 2200 Basic Decimal floating point
Flt.define(:WANG2200,BCDFormat,
  fields:[:significand,13,:exponent,2,:signs,1],
  endiannes: :big_endian,
  bias_mode: :scientific_significand,
  min_exp:-99, max_exp: 99,
  zero_encoded_exp: 0, min_encoded_exp: 0
) { |wang2200|

  # needs special handling because significand and exponent are both stored
  # as sign-magnitude, with both signs combined in a single nibble (decimal digit)
  def wang2200.unpack(v)
    f = unpack_fields_hash(v)
    m = f[:significand]
    e = f[:exponent]
    ss = f[:signs]
    s = (ss==9 || ss==1) ? 9 : 0
    es = (ss==8 || ss==1) ? 9 : 0
    if m==0
      # +-zero
      e = :zero
    elsif @infinite_encoded_exp && e==@infinite_encoded_exp && m==0
      # +-inifinity
      e = :infinity
    elsif @nan_encoded_exp && e==@nan_encoded_exp && m!=0
      # NaN
      e = :nan
    else
      # normalized number
      # reverse exponent nibbles
      e = ("%0#{exponent_digits}d"%e).reverse.to_i
      e = -e if es%2!=0
      e -= (significand_digits-1)
    end
    s = sign_to_unit(s)
    [s,m,e]
  end

  def wang2200.pack(s,m,e)
    s = sign_from_unit(s)
    msb = radix_power(@significand_digits-1)
    es = 0
    if e==:zero
      e = @zero_encoded_exp
      m = 0
    elsif e==:infinity
      e = @infinite_encoded_exp || radix_power(@fields[:exponent])-1
      m = 0
    elsif e==:nan
      e = @infinite_encoded_exp || radix_power(@fields[:exponent])-1
      s = minus_sign_value # ?
      m = radix_power(@significand_digits-2) if m==0
    elsif e==:denormal
      e = @denormal_encoded_exp
    else
      # to do: try to adjust m to keep e in range if out of valid range
      # to do: reduce m and adjust e if m too big

      min_exp = radix_min_exp(:integral_significand)
      if m>0
        while m<msb && e>min_exp
          e -= 1
          m *= radix
        end
      end
      if m<msb && @denormal_encoded_exp
        e = @denormal_encoded_exp
      elsif m==0 # => && @denormal_encoded_exp.nil?
        e = 0
      else
        e += (significand_digits-1)
        if e<0
          e = -e
          es = 9
        else
          es = 0
        end
      end
    end
    ss = (s%2) + (es==0 ? 0 : 8)
    # reverse exponent nibbles
    e = ("%0#{exponent_digits}d"%e).reverse.to_i
    pack_fields_hash signs: ss, significand: m, exponent: e
  end
}


# C51 (C compiler for the Intel 8051) BCD Floating point formats
class C51BCDFormat < BCDFormat # :nodoc:
  def self.exponent_radix
    2
  end
  def self.exponent_digits
    @fields[:exponent_sign]*4-1
  end
  def self.minus_sign_value
    1
  end
  def self.unpack(v)
    f = unpack_fields_hash(v)
    m = f[:significand]
    e_s = f[:exponent_sign]
    exp_bits = exponent_digits
    e = e_s & (2**exp_bits-1)
    s = e_s >> exp_bits
    if m==0
      # +-zero
      e = :zero
    elsif @infinite_encoded_exp && e==@infinite_encoded_exp && m==0
      # +-inifinity
      e = :infinity
    elsif @nan_encoded_exp && e==@nan_encoded_exp && m!=0
      # NaN
      e = :nan
    else
      # normalized number
      e = decode_exponent(e, :integral_significand)
    end
    s = sign_to_unit(s)
    [s,m,e]
  end
  def self.bcd_field?(i)
    @field_meaning[i]==:significand
  end

  def self.pack(s,m,e)
    s = sign_from_unit(s)
    msb = radix_power(@significand_digits-1)
    es = 0
    if e==:zero
      e = @zero_encoded_exp
      m = 0
    elsif e==:infinity
      e = @infinite_encoded_exp || radix_power(@fields[:exponent])-1
      m = 0
    elsif e==:nan
      e = @infinite_encoded_exp || radix_power(@fields[:exponent])-1
      s = minus_sign_value # ?
      m = radix_power(@significand_digits-2) if m==0
    elsif e==:denormal
      e = @denormal_encoded_exp
    else
      # to do: try to adjust m to keep e in range if out of valid range
      # to do: reduce m and adjust e if m too big

      min_exp = radix_min_exp(:integral_significand)
      if m>0
        while m<msb && e>min_exp
          e -= 1
          m *= radix
        end
      end
      if m<msb && @denormal_encoded_exp
        e = @denormal_encoded_exp
      elsif m==0 # => && @denormal_encoded_exp.nil?
        e = 0
      else
        e = encode_exponent(e, :integral_significand)
      end
    end
    exp_bits = exponent_digits
    e_s = e + (s << exp_bits)
    pack_fields_hash significand: m, exponent_sign: e_s
  end
end

Flt.define :C51_BCD_FLOAT, C51BCDFormat,
  fields:[:exponent_sign, 2, :significand,6],
  endiannes: :big_endian,
  bias: 64, bias_mode: :fractional_significand,
  zero_encoded_exp: 0, min_encoded_exp: 0,max_encoded_exp: 127

Flt.define :C51_BCD_DOUBLE,C51BCDFormat,
  fields:[:exponent_sign, 2, :significand,10],
  endiannes: :big_endian,
  bias: 64, bias_mode: :fractional_significand,
  zero_encoded_exp: 0, min_encoded_exp: 0,max_encoded_exp: 127

Flt.define :C51_BCD_LONG_DOUBLE, C51BCDFormat,
  fields:[:exponent_sign, 2, :significand,12],
  endiannes: :big_endian,
  bias: 64, bias_mode: :fractional_significand,
  zero_encoded_exp: 0, min_encoded_exp: 0,max_encoded_exp: 127


# double-double format as used in the PowerPC
Flt.define :IEEE_DOUBLE_DOUBLE, DoubleFormat, half: IEEE_binary64, extra_prec: true



end
